src/plugins/tonegen/stream.c

```diff
@@ -0,0 +1,801 @@
+}
+
+void stream_destroy(struct stream *stream)
+{
+    struct ausrv         *ausrv = stream->ausrv;
+    struct stream        *prev;
+    pa_stream            *pastr;
+    const pa_buffer_attr *battr;
+    struct stream_stat   *stat;
+    pa_operation         *oper;
+#ifdef ENABLE_TRACE
+    struct timeval        tv;
+    uint64_t              stop;
+    double                upt;
+    double                strt;
+    double                dur;
+    double                freq;
+    double                flow;
+    uint32_t              avbuf;
+    uint32_t              avcalc;
+    uint32_t              avcpu;
+    uint32_t              avgap;
+#endif
+
+    TRACE("%s(): destroying stream '%s'", __FUNCTION__, stream->name);
+
+
+    if (stream->killed) {
+        TRACE("%s(): stream is already killed", __FUNCTION__);
+        return;
+    }
+
+#ifdef ENABLE_TRACE
+    gettimeofday(&tv, NULL);
+    stop = (uint64_t)tv.tv_sec * (uint64_t)1000000 + (uint64_t)tv.tv_usec;
+#endif
+
+
+    for (prev=(struct stream *)&ausrv->streams;  prev->next;  prev=prev->next){
+        if (prev->next == stream) {
+            pastr = stream->pastr;
+            battr = pa_stream_get_buffer_attr(pastr);
+            stat  = &stream->stat;
+
+            if (stream->flush)
+                oper = pa_stream_flush(pastr, flush_callback, (void *)stream);
+            else
+                oper = pa_stream_drain(pastr, drain_callback, (void *)stream);
+
+            if (oper == NULL)
+                return;
+
+            pa_operation_unref(oper);
+
+            prev->next = stream->next;
+            stream->next   = NULL;
+            stream->killed = TRUE;
+
+            if (stream->destroy != NULL)
+                stream->destroy(stream->data);
+
+            stream->ausrv  = NULL;
+
+            free(stream->buf.samples);
+
+            pa_stream_set_write_callback(pastr, NULL,NULL);
+
+            if (print_statistics && stat->wrcnt > 0) {
+                if (battr != NULL) {
+                    TRACE("Buffer attributes:\n"
+                          "   maxlength %u\n"
+                          "   tlength   %u\n"
+                          "   prebuf    %u\n"
+                          "   minreq    %u",
+                          battr->maxlength, battr->tlength,
+                          battr->prebuf, battr->minreq);
+                }
+
+#ifdef ENABLE_TRACE
+                upt  = (double)(stop - stream->start) / 1000000.0;
+                strt = (double)(stream->time) / 1000000.0;
+                dur  = (double)(stat->wrtime - stat->firstwr)/1000000.0 + 0.01;
+                freq = (double)stat->wrcnt / dur;
+                flow = (double)stream->bcnt / dur;
+
+                avbuf  = stream->bcnt / stat->wrcnt;
+                avcpu  = (stat->cpucalc / stat->wrcnt) / 1000;
+                avcalc = (uint32_t)(stat->sumcalc/(uint64_t)stat->wrcnt)/1000; 
+                avgap  = (uint32_t)(stat->sumgap /(uint64_t)stat->wrcnt)/1000; 
+#endif
```

Comment:
This `#endif` can probably be placed after the `TRACE()` call below just for good measure?